pkg/bmc/bmc.go

```diff
@@ -0,0 +1,349 @@
+package bmc
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/stmcginnis/gofish"
+	"github.com/stmcginnis/gofish/redfish"
+)
+
+const (
+	defaultTimeOut = 5 * time.Second
+)
+
+var (
+	// DefaultTimeOuts holds the default redfish and ssh timeouts.
+	DefaultTimeOuts = TimeOuts{
+		Redfish: defaultTimeOut,
+		SSH:     defaultTimeOut,
+	}
+)
+
+// User holds the Name and Password for a user (ssh/redfish).
+type User struct {
+	// Name holds the user's name
+	Name string
+	// Password holds the user's password
+	Password string
+}
+
+// TimeOuts holds the configured timeouts for Redfish and SSH acccess.
+type TimeOuts struct {
+	// Redfish timeout for the redfish api access.
+	Redfish time.Duration
+	// SSH timeout for the ssh access.
+	SSH time.Duration
+}
+
+// BMC is the holder struct for BMC access through redfish & ssh.
+type BMC struct {
+	host        string
+	redfishUser User
+	sshUser     User
+	systemIndex int
+
+	timeOuts TimeOuts
+}
+
+// New returns a new BMC struct. The default system index to be used in redfish requests is 0.
+// Use SetSystemIndex to modify it.
+func New(host string, redfishUser, sshUser User, timeOuts TimeOuts) (*BMC, error) {
+	errMsgs := []string{}
+	if host == "" {
+		errMsgs = append(errMsgs, "host is empty")
+	}
+
+	if redfishUser.Name == "" {
+		errMsgs = append(errMsgs, "redfish user's name is empty")
+	}
+
+	if redfishUser.Password == "" {
+		errMsgs = append(errMsgs, "redfish user's password is empty")
+	}
+
+	if sshUser.Name == "" {
+		errMsgs = append(errMsgs, "ssh user's name is empty")
+	}
+
+	if sshUser.Password == "" {
+		errMsgs = append(errMsgs, "ssh user's password is empty")
+	}
+
+	if timeOuts.Redfish == 0 {
+		errMsgs = append(errMsgs, "redfish timeout is 0")
+	}
+
+	if timeOuts.SSH == 0 {
+		errMsgs = append(errMsgs, "ssh timeout is 0")
+	}
+
+	// Build final error msg in case there were some error/s validating the input params.
+	if len(errMsgs) > 0 {
+		errMsg := ""
+
+		for i, msg := range errMsgs {
+			if i != 0 {
+				errMsg += ", "
+			}
+
+			errMsg += msg
+		}
+
+		glog.V(100).Infof("Failed to initialize BMC: %s", errMsg)
+
+		return nil, errors.New(errMsg)
+	}
+
+	glog.V(100).Infof("Initializing new BMC structure with the following params: %s, %v, %v, %v (system index = 0)",
```

Comment:
Please move this logs to the top of this function. It's OK to print log which shows that function is triggered and then fall on error if occur

Reply:
Done in https://github.com/openshift-kni/eco-goinfra/pull/378/commits/afdcff2edbed0d107f30b4104c66db97f47e6bbe